app/src/main/java/org/sil/gatherwords/SessionActivity.java

```diff
@@ -118,18 +128,110 @@ public void save_settings_fab_pressed(View view) {
+    private static class InsertSessionsTask extends AsyncTask<Session, Double, Void> {
         private SessionDao sDAO;
+        private WordDao wordDao;
+        private WeakReference<SessionActivity> sessionActivityReference;
+        private Long sessionID;
+        private int maxProgress;
+        private AppDatabase db;
+        private String wordList;
 
-        InsertSessionsTask(AppDatabase db) {
+        InsertSessionsTask(SessionActivity activity) {
+            sessionActivityReference = new WeakReference<>(activity);
+            db = AppDatabase.get(sessionActivityReference.get().getApplicationContext());
+            wordList = activity.worldListToLoad;
+            wordDao = db.wordDao();
             sDAO = db.sessionDao();
         }
 
         @Override
         protected Void doInBackground(Session... sessions) {
-            sDAO.insertSession(sessions);
+            long[] ids;
+            ids = sDAO.insertSession(sessions);
+
+            if (ids.length > 0) {
+                sessionID = ids[0];
+                insertFromAsset();
+            }
             return null;
         }
+
+        /**
+         * Insert the words from the selected asset or exit if none selected
+         */
+        private void insertFromAsset() {
+            if (wordList.equals("")) {
+                // There was no file selected
+                return;
+            }
+
+            // Begin the transaction
+            db.beginTransaction();
+            try {
+                JSONArray jsonArray = new JSONArray(loadWordList());
+                maxProgress = jsonArray.length();
+
+                // Iterate through each word from the file
+                for (int i = 0; i < jsonArray.length(); i++) {
+                    JSONObject json = jsonArray.getJSONObject(i);
+                    Word word = new Word();
+
+                    // Link the word to the session that was just created
+                    word.sessionId = sessionID;
+                    word.meanings = json.getString("entry");
+
+                    // Insert
+                    wordDao.insertWords(word);
+                }
+
+                // Mark to commit the changes to the DB
+                db.setTransactionSuccessful();
+            } catch (Exception e) {
+                Log.e("InsertSessionsTask", "Exception while inserting words", e);
+            }
+            finally {
+                // Commit or rollback the database
+                db.endTransaction();
+            }
+        }
+
+        /**
+         * Read the selected file
+         *
+         * @return The contents of the file
+         */
+        private String loadWordList() {
+            InputStream is = null;
+            BufferedReader br = null;
+            String file = null;
+            try {
+                is = sessionActivityReference.get().getApplicationContext().getAssets().open("wordLists/" + wordList);
```

Comment:
`WeakReference` may be garbage collected at any time, so this could throw a npe.

Reply:
Then what's the point of using the `WeakReference` type if I can't use it to avoid the possible data leak from storing a `static Context` variable?  
What other options will work?